docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,179 @@
+
+## Considered Options
+
+* Redesign KDS
+
+## Decision Outcome
+
+Chosen option: "Redesign KDS".
+
+## Solution
+
+Currently, we are using one bidirection stream for sending resources from Zone to Global and Global to Zone.
+
+```proto
+service KumaDiscoveryService {
+ rpc StreamKumaResources(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryResponse);
+}
+```
+
+That is hidden behind `MultiplexService`.
+
+``` proto
+service MultiplexService {
+  rpc StreamMessage(stream Message) returns (stream Message);
+}
+```
+
+Zone to Global synchronize resources:
+
+* Dataplane
+* DataplaneInsight
+* ZoneEgress
+* ZoneEgressInsight
+* ZoneIngress*
+* ZoneIngressInsight
+
+`*` ZoneIngress resource is synchronized from Global to Zone and Zone to Global
+
+Global to Zone synchronize resources:
+* CircuitBreaker
+* ExternalService
+* FaultInjection
+* HealthCheck
+* Mesh
+* MeshGateway
+* MeshGatewayRoute
+* ProxyTemplate
+* RateLimit
+* Retry
+* ServiceInsight
+* Timeout
+* TrafficLog
+* TrafficPermission
+* TrafficRoute
+* TrafficTrace
+* VirtualOutbound
+* ZoneIngress*
+* Secret
+* Config
+* MeshAccessLog
+* MeshCircuitBreaker
+* MeshFaultInjection
+* MeshHealthCheck
+* MeshHttpRoute
+* MeshProxyPatch
+* MeshRateLimit
+* MeshRetry
+* MeshTimeout
+* MeshTrace
+* MeshTrafficPermission
+
+A client sends DiscoveryRequest to the server and waits for the change. The server stores the state in a state-of-the-world cache and sends the response whenever there is a change. It works well in the case of Global to Zone synchronization
+but might have scalability issues for synchronization Zone to Global with large deployments. Zone control-plane needs to send all the Dataplane resources to the global whenever there is a change. That might be a bottleneck that can happen in large deployments. We can use [`Incremental xDS`](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol) which is Envoy's implementation of cache that instead of sending all the resources sends only changes. That could reduce the bandwidth usage.
+
+```proto
+service GlobalKDSService {
+ ...
+ rpc ZoneToGlobalSync(stream envoy.service.discovery.v3.DeltaDiscoveryResponse)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryRequest);
+
+ rpc GlobalToZoneSync(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryResponse);
+
+```
+
+In the first iteration, we can introduce only `ZoneToGlobalSync` with delta xDS, and in the future, we could introduce `GlobalToZoneSync` and also use delta xDS. In the code, I've noticed that we are using our copy of the Snapshot cache and the code is a bit outdated. As a part of that task, we should check if we can change to use `go-control-plane` cache implementation or copy and update our current implementation. 
```

Comment:
The way this is written I find it really hard to see what this MADR actually _decides_ and what the alternatives/tradeoffs are. What are we approving? There's a lot of "could"s written in prose.

Reply:
I've rephrased it